From 19f529c1d745545aaaf874956961c11fa077af23 Mon Sep 17 00:00:00 2001
From: Alice Liu <alice.liu@intel.com>
Date: Thu, 10 Nov 2016 20:11:59 +0800
Subject: [PATCH 0386/1240] [REVERTME] dwc3 setup highspeed to USB3.0 on bxtp
 platform

We find USB3.0 dwc3 can't work at superspeed on bxtp platform. We
setup dwc3 highspeed to enable USB3.0 on bxtp platform.

Change-Id: I48df90c39b652f8e2a48598e9e31947e5d68ca17
Signed-off-by: xiao jin <jin.xiao@intel.com>
Signed-off-by: Liu, Alice <alice.liu@intel.com>
Signed-off-by: Greniger, Jacek <jacekx.greniger@intel.com>
(cherry picked from commit 4fc6d0494f0d5f21223c5b3f802a116a757c5893)
---
 drivers/usb/dwc3/gadget.c | 30 ++++++++++++++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/usb/dwc3/gadget.c b/drivers/usb/dwc3/gadget.c
index 5a75c4b..7db3b81 100644
--- a/drivers/usb/dwc3/gadget.c
+++ b/drivers/usb/dwc3/gadget.c
@@ -26,6 +26,7 @@
 #include <linux/io.h>
 #include <linux/list.h>
 #include <linux/dma-mapping.h>
+#include <asm/processor.h>
 
 #include <linux/usb/ch9.h>
 #include <linux/usb/gadget.h>
@@ -1860,6 +1861,17 @@ static void dwc3_gadget_setup_nump(struct dwc3 *dwc)
 	dwc3_writel(dwc->regs, DWC3_DCFG, reg);
 }
 
+static inline bool platform_is_bxtp(void)
+{
+#ifdef CONFIG_X86_64
+	if ((boot_cpu_data.x86_model == 0x5c)
+		&& (boot_cpu_data.x86_mask >= 0x8)
+		&& (boot_cpu_data.x86_mask <= 0xf))
+		return true;
+#endif
+	return false;
+}
+
 static int __dwc3_gadget_start(struct dwc3 *dwc)
 {
 	struct dwc3_ep		*dep;
@@ -2059,10 +2071,24 @@ static void dwc3_gadget_set_speed(struct usb_gadget *g,
 			reg |= DWC3_DCFG_HIGHSPEED;
 			break;
 		case USB_SPEED_SUPER:
-			reg |= DWC3_DCFG_SUPERSPEED;
+			/*
+			 * WORKAROUND: BXTP platform USB3.0 port SS fail,
+			 * We switch SS to HS to enable USB3.0.
+			 */
+			if (platform_is_bxtp())
+				reg |= DWC3_DCFG_HIGHSPEED;
+			else
+				reg |= DWC3_DCFG_SUPERSPEED;
 			break;
 		case USB_SPEED_SUPER_PLUS:
-			reg |= DWC3_DCFG_SUPERSPEED_PLUS;
+			/*
+			 * WORKAROUND: BXTP platform USB3.0 port SS fail,
+			 * We switch SS to HS to enable USB3.0.
+			 */
+			if (platform_is_bxtp())
+				reg |= DWC3_DCFG_HIGHSPEED;
+			else
+				reg |= DWC3_DCFG_SUPERSPEED_PLUS;
 			break;
 		default:
 			dev_err(dwc->dev, "invalid speed (%d)\n", speed);
-- 
2.7.4

